rest/handler_config_database.go

```diff
@@ -0,0 +1,131 @@
+				etagVersion = "0-"
+			}
+			return &dbConfig.DbConfig, etagVersion, nil
+		}
+		return nil, "", nil
+	} else if found, databaseConfig, err := h.server.fetchDatabase(h.db.Name); err != nil {
+		return nil, "", err
+	} else if !found {
+		return nil, "", base.HTTPErrorf(http.StatusNotFound, "database config not found")
+	} else {
+		return &databaseConfig.DbConfig, databaseConfig.Version, nil
+	}
+}
+
+// Updates the database config via a callback function that can modify a `DbConfig`.
+// Note: This always returns a non-nil error; on success it's an HTTPError with status OK.
+// The calling handler method is expected to simply return the result.
+func (h *handler) mutateDbConfig(mutator func(*DbConfig) error) error {
+	h.assertAdminOnly()
+	dbName := h.db.Name
+	validateOIDC := !h.getBoolQuery(paramDisableOIDCValidation)
+
+	if h.server.persistentConfig {
+		// Update persistently-stored config:
+		bucket := h.db.Bucket.GetName()
+		var updatedDbConfig *DatabaseConfig
+		cas, err := h.server.bootstrapContext.connection.UpdateConfig(
+			bucket, h.server.config.Bootstrap.ConfigGroupID,
+			func(rawBucketConfig []byte) (newConfig []byte, err error) {
+				var bucketDbConfig DatabaseConfig
+				if err := base.JSONUnmarshal(rawBucketConfig, &bucketDbConfig); err != nil {
+					return nil, err
+				}
+
+				if h.headerDoesNotMatchEtag("If-Match", bucketDbConfig.Version) {
+					return nil, base.HTTPErrorf(http.StatusPreconditionFailed, "Provided If-Match header does not match current config version")
+				}
+
+				// Now call the mutator function:
+				if err := mutator(&bucketDbConfig.DbConfig); err != nil {
+					return nil, err
+				}
+
+				if err := bucketDbConfig.validate(h.ctx(), validateOIDC); err != nil {
+					return nil, base.HTTPErrorf(http.StatusBadRequest, err.Error())
+				}
+
+				bucketDbConfig.Version, err = GenerateDatabaseConfigVersionID(bucketDbConfig.Version, &bucketDbConfig.DbConfig)
+				if err != nil {
+					return nil, err
+				}
+
+				updatedDbConfig = &bucketDbConfig
+				return base.JSONMarshal(bucketDbConfig)
+			})
+		if err != nil {
+			return err
+		}
+		updatedDbConfig.cas = cas
+
+		dbCreds := h.server.config.DatabaseCredentials[dbName]
+		if err := updatedDbConfig.setup(dbName, h.server.config.Bootstrap, dbCreds); err != nil {
+			return err
+		}
+
+		h.server.lock.Lock()
+		defer h.server.lock.Unlock()
+
+		// TODO: Dynamic update instead of reload
+		if err := h.server._reloadDatabaseWithConfig(*updatedDbConfig, false); err != nil {
+			return err
+		}
+		h.setEtag(updatedDbConfig.Version)
+
+	} else {
+		// Update in-memory config read from JSON file:
+		databaseConfig := *h.server.GetDatabaseConfig(dbName)
+
+		if h.headerDoesNotMatchEtag("If-Match", databaseConfig.Version) {
+			return base.HTTPErrorf(http.StatusPreconditionFailed, "Provided If-Match header does not match current config version")
+		}
+
+		// Now call the mutator function:
+		if err := mutator(&databaseConfig.DbConfig); err != nil {
```

Comment:
Then the `PUT /db/` handler (`handler.handlePutDbConfig()`) has the same bug, because I copied the code from there. (Except it doesn't support Etags at all; I added that.)

I tried fixing the race condition by locking `h.server.lock`, but that causes deadlocks in the handler because it's not a recursive mutex.

Reply:
It looks like this is handled for the persistent config case (UpdateConfig's cas handling is going to ensure the etag checking is valid).  I don't think the potential race in the non-persistent config case (which should be test-only?) requires any additional attention.